### PR TITLE
write permissions for sample PUT, POST, PATCH and DELETE

### PR DESCRIPTION
**Pull request by shriramshankar**

### CODE REVIEW

File: cache/models/samples.js [hunk 1]
```diff
@@ -586,15 +616,31 @@ module.exports = {
    * response. Note: Message body and message code will be updated if provided,
    * else no fields for message body/message code in redis object.
    * @param  {Object} params - Request parameters
+   * @param {String} userName - The user performing the write operation
    * @returns {Promise} - Resolves to a sample object
    */
-  patchSample(params) {
+  patchSample(params, userName) {
     const sampleName = params.key.value;
+    const aspectName = sampleName.split('|')[ONE];
```
Comment (iamigo) [inline cache/models/samples.js]: Is it possible to arrive at this point in the code with an invalid sampleName, e.g. a sampleName with no "|" character, or other ?

File: cache/models/samples.js [hunk 2]
```diff
@@ -737,15 +784,31 @@ module.exports = {
    * if needed. Delete message body and message code fields from hash
    * if not provided because they will be null. Then update sample.
    * @param  {Object} params - Request parameters
+   * @param {String} userName - The user performing the write operation
    * @returns {Promise} - Resolves to a sample object
    */
-  putSample(params) {
+  putSample(params, userName) {
     const sampleName = params.key.value;
+    const aspectName = sampleName.split('|')[ONE];
```
Comment (iamigo) [inline cache/models/samples.js]: same question as above

File: cache/sampleStore.js [hunk 3]
```diff
@@ -167,7 +167,8 @@ function isSampleWritable(aspectModel, aspectName, userName) {
   const options = {};
   options.where = { name: { $iLike: aspectName } };
   return aspectModel.findOne(options)
```
Comment (iamigo) [inline cache/sampleStore.js]: Are we going to the db every time here?

Comment (shriramshankar) [inline cache/sampleStore.js]: Only when the write permission is enabled.

File: cache/models/samples.js (hunk 1 above)
Comment (shriramshankar) [inline cache/models/samples.js]: Yes,  it is possible. I will add the validation.

File: cache/sampleStore.js (hunk 3 above)
Comment (shriramshankar) [inline cache/sampleStore.js]: @iamigo Updated PR and will create a story for caching the related users.

File: cache/models/samples.js [hunk 4]
```diff
@@ -523,14 +538,37 @@ module.exports = {
...
+    if (sampAspArr.length < TWO) {
+      throw new redisErrors.ResourceNotFoundError({
+        explanation: 'Incorrect sample name.',
+      });
+    }
+
+    const aspectName = sampAspArr[ONE];
     let currSampObj;
     let aspectObj;
+    const hasWritePerm = featureToggles
```
Comment (iamigo) [inline cache/models/samples.js]: This same line is repeated multiple times--make it a fn?

**Comment by shriramshankar** [comment]
PR updated. Tested these changes manually also, to make sure they work.

**Comment by pallavi2209** [comment]
Sorry, I missed the review request. Looking at it now.

**Comment by pallavi2209** [comment]
The changes look good. Just one thing to confirm. We do not enforce write permission on creating sample. Correct?

**Comment by iamigo** [comment]
> We do not enforce write permission on creating sample. Correct?

We are supposed to be enforcing based on Aspect's writers. @shriramshankar

**Comment by shriramshankar** [comment]
@iamigo @pallavi2209 PR updated. Perms for POST sample needed to be done for both db and cache layer.

**Comment by shriramshankar** [comment]
@iamigo The changes for upsert/bulkUpsert went in some time back and the tests for that went in along with the changes.   

FYI: few minutes I made a few changes to correct some typos.

**Comment by iamigo** [comment]
@shriramshankar test failure?

  1) redis: aspect: create:  "after each" hook: forceDelete for "when name changes, the sampleStore refelct this change":
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

**Comment by shriramshankar** [comment]
@iamigo Resolved the conflict. The test failure was resolved once I restarted the build again.

**Pull request merged**